File: app/src/main/java/com/mytlogos/enterprise/viewmodel/TocEpisodeViewModel.java
package com.mytlogos.enterprise.viewmodel;

import android.app.Application;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;
import androidx.paging.PagedList;

import com.mytlogos.enterprise.model.TocEpisode;
import com.mytlogos.enterprise.tools.Sortings;
import com.mytlogos.enterprise.ui.ActionCount;

import java.io.IOException;

public class TocEpisodeViewModel extends FilterableViewModel implements SortableViewModel {
    private MutableLiveData<SortFilter> sortFilterLiveData = new MutableLiveData<>();
    private LiveData<PagedList<TocEpisode>> repositoryToc;

    public TocEpisodeViewModel(Application application) {
        super(application);
        this.resetFilter();
    }

    @Override
    public void resetFilter() {
        this.sortFilterLiveData.setValue(new Builder(null).createSortFilter());
    }

    @Override
    public void setSort(Sortings sort) {
        SortFilter value = this.sortFilterLiveData.getValue();
        this.sortFilterLiveData.setValue(new Builder(value).setSortings(sort).createSortFilter());
    }

    public Sortings getSort() {
        SortFilter value = this.sortFilterLiveData.getValue();
        return value == null ? Sortings.INDEX_DESC : value.sortings;
    }

    public void setReadFilter(byte readFilter) {
        SortFilter value = this.sortFilterLiveData.getValue();
        this.sortFilterLiveData.setValue(new Builder(value).setRead(readFilter).createSortFilter());
    }

    public void setSavedFilter(byte savedFilter) {
        SortFilter value = this.sortFilterLiveData.getValue();
        this.sortFilterLiveData.setValue(new Builder(value).setSaved(savedFilter).createSortFilter());
    }

    public byte getSavedFilter() {
        SortFilter value = this.sortFilterLiveData.getValue();
        return value == null ? -1 : value.saved;
    }

    public byte getReadFilter() {
        SortFilter value = this.sortFilterLiveData.getValue();
        return value == null ? -1 : value.read;
    }

    public LiveData<PagedList<TocEpisode>> getToc(int mediumId) {
        if (this.repositoryToc == null) {
            this.repositoryToc = Transformations.switchMap(
                    sortFilterLiveData,
                    input -> repository.getToc(mediumId, input.sortings, input.read, input.saved)
            );
        }
        return this.repositoryToc;
    }

    public void deleteLocalEpisode(int episodeId, double combiIndex, ActionCount count, int mediumId) throws IOException {
        switch (count) {
            case ALL:
                repository.deleteAllLocalEpisodes(mediumId, this.getApplication());
                break;
            case CURRENT:
                repository.deleteLocalEpisode(episodeId, mediumId, this.getApplication());
                break;
            case CURRENT_AND_ONWARDS:
                repository.deleteLocalEpisodesWithHigherIndex(combiIndex, mediumId, this.getApplication());
                break;
            case CURRENT_AND_PREVIOUSLY:
                repository.deleteLocalEpisodesWithLowerIndex(combiIndex, mediumId, this.getApplication());
                break;
        }
    }

    public void updateRead(int episodeId, double combiIndex, ActionCount count, int mediumId, boolean read) throws Exception {
        switch (count) {
            case ALL:
                repository.updateAllRead(mediumId, read);
                break;
            case CURRENT:
                repository.updateRead(episodeId, read);
                break;
            case CURRENT_AND_ONWARDS:
                repository.updateReadWithHigherIndex(combiIndex, read, mediumId);
                break;
            case CURRENT_AND_PREVIOUSLY:
                repository.updateReadWithLowerIndex(combiIndex, read, mediumId);
                break;
        }
    }

    public void reload(int episodeId, double combiIndex, ActionCount count, int mediumId) throws Exception {
        switch (count) {
            case ALL:
                repository.reloadAll(mediumId);
                break;
            case CURRENT:
                repository.reloadSingle(episodeId);
                break;
            case CURRENT_AND_ONWARDS:
                repository.reloadHigherIndex(combiIndex, mediumId);
                break;
            case CURRENT_AND_PREVIOUSLY:
                repository.reloadLowerIndex(combiIndex, mediumId);
                break;
        }
    }

    public void download(int episodeId, double combiIndex, ActionCount count, int mediumId) {
        switch (count) {
            case ALL:
                repository.downloadAll(mediumId, this.getApplication());
                break;
            case CURRENT:
                repository.downloadSingle(episodeId, mediumId, this.getApplication());
                break;
            case CURRENT_AND_ONWARDS:
                repository.downloadHigherIndex(combiIndex, mediumId, this.getApplication());
                break;
            case CURRENT_AND_PREVIOUSLY:
                repository.downloadLowerIndex(combiIndex, mediumId, this.getApplication());
                break;
        }
    }

    private static class Builder {

        private Sortings sortings;
        private byte read;
        private byte saved;

        Builder(SortFilter sortFilter) {
            if (sortFilter == null) {
                sortings = Sortings.INDEX_DESC;
                read = -1;
                saved = -1;
            } else {
                sortings = sortFilter.sortings;
                read = sortFilter.read;
                saved = sortFilter.saved;
            }
        }

        public Builder setSortings(Sortings sortings) {
            this.sortings = sortings;
            return this;
        }

        public Builder setRead(byte read) {
            if (read < -1) {
                read = -1;
            } else if (read > 1) {
                read = 1;
            }
            this.read = read;
            return this;
        }

        public Builder setSaved(byte saved) {
            if (saved < -1) {
                saved = -1;
            } else if (saved > 1) {
                saved = 1;
            }
            this.saved = saved;
            return this;
        }

        public SortFilter createSortFilter() {
            return new SortFilter(sortings, read, saved);
        }
    }

    private static class SortFilter {
        private final Sortings sortings;
        /**
         * -1 for ignore
         * 0 for false
         * 1 for true
         */
        private final byte read;
        /**
         * -1 for ignore
         * 0 for false
         * 1 for true
         */
        private final byte saved;

        private SortFilter(Sortings sortings, byte read, byte saved) {
            this.sortings = sortings;
            this.read = read;
            this.saved = saved;
        }
    }
}

File: app/src/main/java/com/mytlogos/enterprise/viewmodel/TocEpisodeViewModel.kt
package com.mytlogos.enterprise.viewmodel

import android.app.Application
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.asFlow
import androidx.paging.PagingData
import com.mytlogos.enterprise.background.repository.EpisodeRepository
import com.mytlogos.enterprise.model.TocEpisode
import com.mytlogos.enterprise.tools.Sortings
import com.mytlogos.enterprise.ui.ActionCount
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.flatMapLatest
import java.io.IOException
import kotlin.math.max
import kotlin.math.min

@Suppress("BlockingMethodInNonBlockingContext")
class TocEpisodeViewModel(application: Application) :
    FilterableViewModel(application),
    SortableViewModel {
    private val sortFilterLiveData = MutableLiveData(SortFilter())
    private val mediumId: MutableStateFlow<Int> = MutableStateFlow(-1)
    private val episodeRepository: EpisodeRepository by lazy {
        EpisodeRepository.getInstance(application)
    }

    @ExperimentalCoroutinesApi
    val toc: Flow<PagingData<TocEpisode>> by lazy {
        mediumId.combine(sortFilterLiveData.asFlow()) { id: Int, filter: SortFilter ->
            episodeRepository.getToc(
                id,
                filter.sortings,
                filter.read,
                filter.saved
            )
        }.flatMapLatest { it }
    }

    fun setMediumId(mediumId: Int) {
        this.mediumId.value = mediumId
    }

    override fun resetFilter() {
        sortFilterLiveData.value = SortFilter()
    }

    fun getSort(): Sortings {
        val value = sortFilterLiveData.value
        return value?.sortings ?: Sortings.INDEX_DESC
    }

    override fun setSort(sort: Sortings) {
        val value = sortFilterLiveData.value
        sortFilterLiveData.value = value!!.copy(sortings = sort)
    }

    var savedFilter: Byte
        get() {
            val value = sortFilterLiveData.value
            return value?.saved ?: -1
        }
        set(savedFilter) {
            val value = sortFilterLiveData.value
            sortFilterLiveData.value = value!!.copy(saved = savedFilter)
        }
    var readFilter: Byte
        get() {
            val value = sortFilterLiveData.value
            return value?.read ?: -1
        }
        set(readFilter) {
            val value = sortFilterLiveData.value
            sortFilterLiveData.value = value!!.copy(read = readFilter)
        }

    @Throws(IOException::class)
    suspend fun deleteLocalEpisode(
        episodeIds: Set<Int>,
        combiIndices: List<Double>,
        count: ActionCount?,
        mediumId: Int,
    ) {
        when (count) {
            ActionCount.ALL -> episodeRepository.deleteAllLocalEpisodes(
                mediumId,
                getApplication()
            )
            ActionCount.CURRENT -> episodeRepository.deleteLocalEpisodes(
                episodeIds,
                mediumId,
                getApplication()
            )
            ActionCount.CURRENT_AND_ONWARDS -> {
                val lowest = getLowest(combiIndices)
                episodeRepository.deleteLocalEpisodesWithHigherIndex(
                    lowest,
                    mediumId,
                    getApplication()
                )
            }
            ActionCount.CURRENT_AND_PREVIOUSLY -> {
                val highest = getHighest(combiIndices)
                episodeRepository.deleteLocalEpisodesWithLowerIndex(
                    highest,
                    mediumId,
                    getApplication()
                )
            }
        }
    }

    @Throws(Exception::class)
    suspend fun updateRead(
        episodeIds: Set<Int>,
        combiIndices: List<Double>,
        count: ActionCount?,
        mediumId: Int,
        read: Boolean,
    ) {
        when (count) {
            ActionCount.ALL -> episodeRepository.updateAllRead(mediumId, read)
            ActionCount.CURRENT -> episodeRepository.updateRead(episodeIds, read)
            ActionCount.CURRENT_AND_ONWARDS -> {
                val lowest = getLowest(combiIndices)
                episodeRepository.updateReadWithHigherIndex(lowest, read, mediumId)
            }
            ActionCount.CURRENT_AND_PREVIOUSLY -> {
                val highest = getHighest(combiIndices)
                episodeRepository.updateReadWithLowerIndex(highest, read, mediumId)
            }
        }
    }

    @Throws(Exception::class)
    suspend fun reload(
        episodeIds: Set<Int>,
        combiIndices: List<Double>,
        count: ActionCount?,
        mediumId: Int,
    ) {
        when (count) {
            ActionCount.ALL -> episodeRepository.reloadAll(mediumId)
            ActionCount.CURRENT -> episodeRepository.reload(episodeIds)
            ActionCount.CURRENT_AND_ONWARDS -> {
                val lowest = getLowest(combiIndices)
                episodeRepository.reloadHigherIndex(lowest, mediumId)
            }
            ActionCount.CURRENT_AND_PREVIOUSLY -> {
                val highest = getHighest(combiIndices)
                episodeRepository.reloadLowerIndex(highest, mediumId)
            }
        }
    }

    suspend fun download(
        episodeIds: Set<Int>,
        combiIndices: List<Double>,
        count: ActionCount?,
        mediumId: Int,
    ) {
        when (count) {
            ActionCount.ALL -> episodeRepository.downloadAll(mediumId, getApplication())
            ActionCount.CURRENT -> episodeRepository.download(episodeIds,
                mediumId,
                getApplication())
            ActionCount.CURRENT_AND_ONWARDS -> {
                val lowest = getLowest(combiIndices)
                episodeRepository.downloadHigherIndex(lowest, mediumId, getApplication())
            }
            ActionCount.CURRENT_AND_PREVIOUSLY -> {
                val highest = getHighest(combiIndices)
                episodeRepository.downloadLowerIndex(highest, mediumId, getApplication())
            }
        }
    }

    private fun getLowest(combiIndices: List<Double>): Double {
        return combiIndices.fold(Int.MAX_VALUE.toDouble(), ::min)
    }

    private fun getHighest(combiIndices: List<Double>): Double {
        return combiIndices.fold(Int.MIN_VALUE.toDouble(), ::max)
    }

    private data class SortFilter(
        val sortings: Sortings = Sortings.INDEX_DESC,
        /**
         * -1 for ignore
         * 0 for false
         * 1 for true
         */
        val read: Byte = -1,
        /**
         * -1 for ignore
         * 0 for false
         * 1 for true
         */
        val saved: Byte = -1,
    )

    init {
        resetFilter()
    }
}